File: Graph/Graph/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Preorder and Inorder Traversal.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int preIndex;
    unordered_map <int,int> inorderIndexMap;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
       preIndex = 0;
       for(int i = 0; i < inorder.size(); i++)
           inorderIndexMap[ inorder[i] ] = i;
      // for(int i = 0; i < inorder.size(); i++)
         //  inorderIndexMap[ inorder[i] ] = 0;
        return arrayToTree(preorder, 0, inorder.size() - 1);
    }
    TreeNode* arrayToTree(vector<int>& preorder, int left, int right){
        if(left > right ) return NULL;
        TreeNode* root = new TreeNode(preorder[preIndex]);
        preIndex++;
        root->left = arrayToTree(preorder,left, inorderIndexMap[root->val] - 1);
        root->right =arrayToTree(preorder, inorderIndexMap[root->val] + 1, right);
        return root;
    }
};

File: Graph/Graph/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length == 0 || inorder.length == 0) return null;
        return helper(0, 0 , inorder.length, preorder, inorder);
    }
   public TreeNode helper(int preStart, int inStart, int inEnd , int[] preorder, int[] inorder){
       if(preStart > preorder.length-1 || inStart > inEnd ) return null;
       TreeNode root = new TreeNode(preorder[preStart]);
       int mid = 0;
       for(int i = 0; i < inorder.length; i++) {
           if( inorder[i] == root.val ){
               mid = i; break;
           }
       }
       root.left = helper(preStart + 1, inStart, mid - 1, preorder, inorder);
       root.right = helper(preStart + mid - inStart+1, mid+1, inEnd, preorder, inorder);
       return root;
   } 
}
